### PR TITLE
Improve Manifset creation

### LINKED ISSUE #2295: Gradle does not preserve .jar manifest attribute order
.jar manifest attributes are written in random order, instead of the one provided in the buildscript.´
See this old issue: [GRADLE-223](https://issues.gradle.org/browse/GRADLE-223)

### Expected Behavior
When writing manifest files, the attributes should be written in the same order as they are specified in the build script.

### Current Behavior
Attributes are written in random order.

### Context
Manifest files are hard to read, and in some corner cases attribute order is actually important.

### Steps to Reproduce (for bugs)
Take a java based multi-project build and use the following:
```
allprojects {
    afterEvaluate {
        jar {
            manifest {
                attributes(
                        'Build-Date': new Date().format("yyyy-MM-dd"),
                        'Build-Time': new Date().format("HH:mm:ssZ"),
                        'Code-Revision': '1.0',
                        'Class-Path': configurations.compile.collect { it.getName() }.join(' ')
                )
            }
        }
    }
}
```

Manifest will have these entries mixed in random order, e.g.

```
Build-Date: 2017-06-14
Code-Revision: 1.0
Build-Time: 15:11:01+0200
Class-Path: xyz.jar
```

### Your Environment
windows 10
java 1.8 jdk
gradle 3.5

### PR DESCRIPTION
**Pull request by vlsi**
Print entries in the proper order (Java8 uses HashMap), and split UTF-8 properly

Note: this is WIP, however comments are welcome.

Fixes #2295, #5225

### Context

Java's `Manifest` implementation still corrupts UTF-8 chars, and Java 8 uses `HashMap` for attributes, thus it shuffles entries.

### Implemented changes
- [x] ManifestWriter/ManifestPrinter: those classes generate manifest and wrap lines accordingly
- [x] Tests for ManifestWriter/ManifestPrinter: 丈, 😃, and नि
- [x] ManifestReader: Java's Manifest can't be used (at least for Java 8) because it shuffles entries.
- [x] Tests for ManifestReader: compare values that are received with ManifestReader and with Java's Manifest
- [x] Roundtrip tests for ManifestWriter-ManifestReader: ensure the values are the same, and the order of entries is intact.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective. **Vladimir**: I don't think new integration tests are needed here.
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes. **Vladimir**: this does not look like a user-visible change, except "bug fixed" in the release notes. However, it would be really nice to add something like "Gradle now supports emojis in META-INF/MANIFEST.MF" :)
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by vlsi** [comment]
An "interesting" bit might be if it is legal to break things like नि in two halves.
Technically speaking, this glyph is composed of two `char` values, however they are not a surrogate pair. That is if you copy and paste them individually, they work, and they merge together when they happen to be placed next to each other.

On the one hand splitting that is kind of "ok-ish", because they would merge to the final character on `manifest read`.
However editing that `split file` might become a bit complicated.

Then, there's [html regex stackoverflow](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454) that suggests cases like Rege̿̔̉x or even rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆

It turns out that the "letters" of rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆ are as follows:

[0..1), length: 1, bytes: 1: r

[1..6), length: 5, bytes: 9: è̑ͧ̌

[6..8), length: 2, bytes: 3: aͨ

[8..17), length: 9, bytes: 17: l̘̝̙̃ͤ͂̾̆

The very last `L` is represented by 9 `chars`, and it consumes 17 bytes in UTF-8 encoding.
Apparently it is not that hard type a glyph that would consume more than 72 bytes (which are allowed per the manifest spec).

Any thoughts if that should be considered?

@floscher, do you think if rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆ in META-INF/MANIFEST.MF is crazy enough?

PS. In case you wondered, the current code would produce (as you see, the combined characters merge right into 1, 22, 333, 4454, 55555, and 666666 which are just manifest keys):

1: rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ
 ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̃ͤ
 ̘̝̙͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̃ͤ͂̾̆
 ̙rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ
 ̌aͨl̘̝̙̃ͤ͂̾̆
22: rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ
 ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̃ͤ
 ̘̝̙͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̃ͤ͂̾̆
 ̙rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ
 ̌aͨl̘̝̙̃ͤ͂̾̆
333: rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑
 ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̃
 ̘̝̙ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̃ͤ͂̾̆
 ̝̙rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑
 ͧ̌aͨl̘̝̙̃ͤ͂̾̆
4444: rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑
 ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̃
 ̘̝̙ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̃ͤ͂̾̆
 ̝̙rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑
 ͧ̌aͨl̘̝̙̃ͤ͂̾̆
55555: rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè
 ̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl
 ̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̃ͤ͂̾̆
 ̘̝̙rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè
 ̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆
666666: rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè
 ̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl
 ̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̃ͤ͂̾̆
 ̘̝̙rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆rè
 ̑ͧ̌aͨl̘̝̙̃ͤ͂̾̆

**Comment by vlsi** [comment]
PS. I tried to add tests via `spock-genesis`, however it does not support unicode much yet: https://github.com/Bijnagte/spock-genesis/issues/40

**Comment by vlsi** [comment]
It looks like the PR is feature complete.

`:platformJvm:check` passes for me.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java [hunk 1]
```diff
@@ -48,35 +80,40 @@ public boolean containsValue(Object value) {
...
+            return attributes.get(keyFor((String) key));
+        }
         return attributes.get(key);
     }
 
     @Override
     public Object put(String key, Object value) {
         if (key == null) {
-            throw new ManifestException("The key of a manifest attribute must not be null.");
+            // NPE is the specified exception type (see Map#put javadoc)
```
Comment (lacasseio) [inline subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: ❌ Let's remove the comment here, as noted, we should comply with the `Map` contract.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java [hunk 2]
```diff
@@ -48,35 +80,40 @@ public boolean containsValue(Object value) {
...
+            throw new NullPointerException("The key of a manifest attribute must not be null.");
         }
         if (value == null) {
-            throw new ManifestException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+            // NPE is the specified exception type (see Map#put javadoc)
```
Comment (lacasseio) [inline subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: ❌ Let's remove the comment here, as noted, we should comply with the `Map` contract.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java [hunk 3]
```diff
@@ -48,35 +80,40 @@ public boolean containsValue(Object value) {
...
+            throw new NullPointerException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+        }
+        Name name;
+        if (key.length() > 70 || key.length() * 3 > 70 && Utf8.encodedLength(key) > 70) {
+            throw new ManifestException(String.format("The Key=%s violates the Manifest spec: key should not exceed 70 bytes, actual length is %d bytes in UTF-8", key, Utf8.encodedLength(key)));
```
Comment (lacasseio) [inline subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: ❓  This may be a breaking change. What is the current behaviour? If it's obviously broken when a user doesn't respect those limits (fails at a later stage), then we could start throwing an exception. In the case where nothing obvious breaks, then we should use a `DeprecationLogger` to nag about this becoming an error in the next Gradle major version.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java [hunk 4]
```diff
@@ -48,35 +80,40 @@ public boolean containsValue(Object value) {
...
+            // NPE is the specified exception type (see Map#put javadoc)
+            throw new NullPointerException("The key of a manifest attribute must not be null.");
         }
         if (value == null) {
-            throw new ManifestException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+            // NPE is the specified exception type (see Map#put javadoc)
+            throw new NullPointerException(String.format("The value of a manifest attribute must not be null (Key=%s).", key));
+        }
+        Name name;
+        if (key.length() > 70 || key.length() * 3 > 70 && Utf8.encodedLength(key) > 70) {
```
Comment (lacasseio) [inline subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: ❌ According to https://docs.oracle.com/javase/7/docs/api/java/util/jar/Attributes.Name.html the name is said to be non-null (which we already check for above), not longer than 70 chars (which we are now checking) and contains only ASCII ([0-9a-zA-Z_-]) chars. The check for `Utf8.encodedLength` would actually be wrong. We should make sure we respect the specs and we can look to see if `new java.util.jar.Attributes.Name(key)` below was doing the right checks so we don't need to implement anything.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java [hunk 5]
```diff
@@ -48,35 +80,40 @@ public boolean containsValue(Object value) {
...
+            throw new ManifestException(String.format("The Key=%s violates the Manifest spec: key should not exceed 70 bytes, actual length is %d bytes in UTF-8", key, Utf8.encodedLength(key)));
         }
         try {
-            new java.util.jar.Attributes.Name(key);
-        } catch (IllegalArgumentException e) {
-            throw new ManifestException(String.format("The Key=%s violates the Manifest spec!", key));
+            name = KNOWN_NAMES.getUnchecked(key);
```
Comment (lacasseio) [inline subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: Great idea, however, in most cases we prefer to prove it's a performance bottleneck before optimizing with caching. I fear this may be premature optimization.
❓ I wonder if this may affect performance. I'm mostly concern about the staticness of the structure where it may keep objects around for longer than required. However, I feel like this may only be cached per build so it may be improved or on par with what we currently have. If I'm not mistaking, this should be in the "per-build" classloader and may be freed after each build. We should confirm we are not worsening the memory pressure of a build.

File: gradle/testDependencies.gradle [hunk 6]
```diff
@@ -36,4 +36,8 @@ testLibraries.mockito_kotlin = 'com.nhaarman:mockito-kotlin:1.6.0'
 testLibraries.mockito_kotlin2 = 'com.nhaarman.mockitokotlin2:mockito-kotlin:2.1.0'
 testLibraries.jackson_kotlin = 'com.fasterxml.jackson.module:jackson-module-kotlin:2.9.2'
 testLibraries.archunit = "com.tngtech.archunit:archunit:0.11.0"
-testLibraries.archunit_junit4 = "com.tngtech.archunit:archunit-junit4:0.11.0"
\ No newline at end of file
+testLibraries.archunit_junit4 = "com.tngtech.archunit:archunit-junit4:0.11.0"
+testLibraries.guava_testlib = 'com.google.guava:guava-testlib:' + libraries.guava.version
+testLibraries.quickcheck_core = 'com.pholser:junit-quickcheck-core:0.9'
+testLibraries.quickcheck_generators = 'com.pholser:junit-quickcheck-generators:0.9'
+testLibraries.bricks_automation = 'dk.brics:automaton:1.12-1'
```
Comment (lacasseio) [inline gradle/testDependencies.gradle]: The tests are incredibly detailed which is really awesome. However, I'm hesitant by the new libraries they require. Could the same kind of coverage be achieved differently? //cc @big-guy

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java [hunk 7]
```diff
@@ -38,6 +67,9 @@ public boolean isEmpty() {
 
     @Override
     public boolean containsKey(Object key) {
+        if (key instanceof String) {
+            return attributes.containsKey(keyFor((String) key));
+        }
```
Comment (lacasseio) [inline subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: ❌ The addition of these checks may actually cause unexpected behaviour if a user would use, previously, invalid keys. They would now be checked by the constructor of `Attributes.Name` which may throw an exception. We will need to consider this either via deprecation warning or accepting the potential breaking change. As I mentioned previously, if it was fundamentally broken before, then we could easily accept the breaking change.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/ManifestInputStream.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+package org.gradle.api.java.archives.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Reads manifest file, and automatically removes value wraps.
+ * Newlines are normalized to LF (for simplified processing later).
+ */
+public class ManifestInputStream extends InputStream {
```
Comment (lacasseio) [inline subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/ManifestInputStream.java]: ❓ I assume these stream classes were extracted from the other classes. I can see how it improve testability and readability. However, I'm concerned by the additional complexity of, what appears to be, handwritten manifest parser. Is there an implementation that already exists that we could rely on?

File: subprojects/platform-jvm/src/test/groovy/org/gradle/api/java/archives/internal/DefaultAttributesTest.groovy [hunk 9]
```diff
@@ -46,22 +46,32 @@ class DefaultAttributesTest extends Specification {
...
+    }
+
+    def caseSensitivity() {
+        when:
+        attributes.put("Hello", "world")
+        attributes.put("hello", "keys are case-insensitive")
+
+        then:
+        attributes.get("Hello") == "keys are case-insensitive"
+        attributes.get("hello") == "keys are case-insensitive"
```
Comment (lacasseio) [inline subprojects/platform-jvm/src/test/groovy/org/gradle/api/java/archives/internal/DefaultAttributesTest.groovy]: ❌ This seems like a potential breakage that we may need to account for. Either accept it if it's fundamentally broken or nag about the change of behaviour in the next major release.

File: gradle/testDependencies.gradle (hunk 6 above)
Comment (vlsi) [reply gradle/testDependencies.gradle]: @lacasseio , I support your view, however, my answer to the same question was

1) `guava:guava-testlib` is the only available library that **automatically** tests `Map` implementations. I'm afraid it can't really be replaced without writing **lots** of tests manually.

2) `quickcheck` can be replaced with hand-baked `random-driven` test, however, `quickcheck` is known at StackOverflow, and `quickcheck` implements automatic shrinking of the reproducers. That helps a lot for debugging of the failures => it simplifies the maintenance.

3) `dk.brics:automaton` is a "well known" automata-driven regexp engine. It enables to write test cases like "I want a sting that conforms to this regexp". I gues that simplifies the understanding of the tests. Of course I can create hand-rolled implementation for the specific case in question, however I think automation-based approach is simpler.

Note: I would like to incorporate bricks_automation-driven regexp generator right into quickcheck, however it is not clear what the release schedule for quickcheck is, so I just wen ahead and added one (or two) classes right into the test code.

Comment (vlsi) [reply gradle/testDependencies.gradle]: At the end of the day, `testLibraries` do not sneak to the resulting binary.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java (hunk 7 above)
Comment (vlsi) [reply subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: manifest attributes have to be case-insensitive.
Previous implementation did not consider that, so users could get "not found" response when they seek for the key in "wrong casing".

New implementation:
1) Supports case-insensitive search
2) It fails if input string does not conform to the manifest spec. Well, it did not fail previously (it just returned "not found" response), however, I don't think it is bad to throw exceptions on invalid input keys.

I'm fine to just ignore the exceptions though, however, I think no-one would really use `containsKey` API. The key feature here is "enable case-insensitive search".

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java (hunk 1 above)
Comment (vlsi) [reply subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: I can remove it, however, I left it for those who would discover the code and ask the classic "WTF?" question. It might prevent back-and-forth changes.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java [hunk 10]
```diff
@@ -48,35 +80,40 @@ public boolean containsValue(Object value) {
...
         return attributes.get(key);
     }
 
     @Override
     public Object put(String key, Object value) {
         if (key == null) {
-            throw new ManifestException("The key of a manifest attribute must not be null.");
+            // NPE is the specified exception type (see Map#put javadoc)
+            throw new NullPointerException("The key of a manifest attribute must not be null.");
```
[the remark this answers is not in the document]
Comment (vlsi) [reply subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: PS. That comes from guava-testlib.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java (hunk 4 above)
Comment (vlsi) [reply subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: > The check for Utf8.encodedLength would actually be wrong

@lacasseio , can you please clarify?

Here's the specification:
https://docs.oracle.com/javase/7/docs/technotes/guides/jar/jar.html#Manifest_Specification

> No line may be longer than 72 bytes (not characters), in its UTF8-encoded form

> Because header names cannot be continued, the maximum length of a header name is 70 bytes (there must be a colon and a SPACE after the name).

The specification says that the maximum allowed key is 70 **bytes**
`Attributes.Name` uses `string#length`, thus it **fails** to comply with the specification.

Do you really suggest there's a bug in the specification itself?
Can you please provide the link to the **proper** specification?

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java (hunk 3 above)
Comment (vlsi) [reply subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: Frankly speaking, in 99.42% of the cases, everybody uses `plain old ascii` for **keys**.
Thus string length and UTF-8 length are basically the same.
However, the manifest spec forbids to split manifest keys, so manifest parsers might easily fail to parse those split key names.

Even though it is a breaking change, I don't think it is really important to somehow support long manifest keys.


However, if you feel strongly I can add `DeprecationLogger` (or whatever), however, I would need a pointer to the similar feature. Gradle's codebase is huge, and I'm not familiar with approaches to "feature flags", "deprecation warnings", etc, etc.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java (hunk 5 above)
Comment (vlsi) [reply subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: Frankly speaking, I don't have lots of Manifests to test on.
I just peeked on OpenJDK implementation, and I see there's a cache there.

Then I don't like the idea that `Attributes.Name` performs `name.intern()`, so I added the cache.

> worsening the memory pressure of a build

There's limit of 1000 entries maximum. It can be lowered, however I hope people do not use randomized key names for manifests.
They would probably use 10-40 different key names maximum.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/ManifestInputStream.java (hunk 8 above)
Comment (vlsi) [reply subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/ManifestInputStream.java]: I do not know if such an implementation exists.

Java's implementation is broken due to the use of `HashMap`.
Ant Manifest class doesn't support split multi-bytes characters.

I agree it is sad to have yet another manifest parser, however, the resulting implementation seems to be more-or-less clean.

I don't think it would be way better if I publish manifest parser under `com.github.vlsi:manifest-parser` coordinates. Would it?

File: subprojects/platform-jvm/src/test/groovy/org/gradle/api/java/archives/internal/DefaultAttributesTest.groovy (hunk 9 above)
Comment (vlsi) [reply subprojects/platform-jvm/src/test/groovy/org/gradle/api/java/archives/internal/DefaultAttributesTest.groovy]: @lacasseio , `java.util.jar.Attributes` javadoc says:

>Valid attribute names are case-insensitive

Gradle's `DefaultAttributes` pretend to `implement java.util.jar.Attributes`, so it should adhere to the interface contract.

File: subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java (hunk 4 above)
Comment (vlsi) [reply subprojects/platform-jvm/src/main/java/org/gradle/api/java/archives/internal/DefaultAttributes.java]: Ah, my bad here. The keys are limited to alpha+digit+_-, so string length and UTF-8 length is the same for that.

I guess I can drop Utf8.encodedLength(key) checks here.

File: subprojects/platform-jvm/src/test/groovy/org/gradle/api/java/archives/internal/DefaultAttributesTest.groovy (hunk 9 above)
Comment (vlsi) [reply subprojects/platform-jvm/src/test/groovy/org/gradle/api/java/archives/internal/DefaultAttributesTest.groovy]: Oh. I missed that `DefaultAttributes` implements `org.gradle.api.java.archives.Attributes` which is not really connected with `java.util.jar.Attributes`.

However, the manifest specification says that `Attribute names are case insensitive`, so Gradle should probably adhere to that somehow. Looking forward to hearing from you on your decision if Gradle wants to support case-insensitive attribute names.

**Comment by vlsi** [comment]
@lacasseio , thanks for the review.

I've came across the specification one more time, and I find the following:

>https://docs.oracle.com/javase/7/docs/technotes/guides/jar/jar.html#Notes_on_Manifest_and_Signature_Files
>Before parsing:
>If the last character of the file is an EOF character (code 26), the EOF is treated as whitespace. Two newlines are appended (one for editors that don't put a newline at the end of the last line, and one so that the grammar doesn't have to special-case the last entry, which may not have a blank line after it).

I can probably add a test with "file ending with code 26".

Then, I don't get what is "Two newlines are appended". It would be great if you could provide a hint or something on the meaning behind the words.

Just in case: current tests verify that new implementation is compatible with Java's and with Ant's.
However, there might be "manually crafted manifests" with "Two newlines are appended" feature. Then those files might fail to parse properly. I don't know how to test that since I don't quite understand the intention behind "two newlines are appended"

**Comment by vlsi** [comment]
So far:
* I've rebased the PR on top of the current master
* implemented&tested "ctrl-z byte removal" when it is the last byte in the stream (manifest spec says the byte should be ignored then)
* added "Gradle Manifest -> Ant Manifest" and "Ant -> Gradle" tests
* added `ManifestPrinterTest.provider in values` (an extra test for fb60acb0bd127b5c6becdbbb9662e1a6f68c0e64 )

**Comment by lacasseio** [comment]
Sorry @vlsi for the late reply. We really appreciate the effort you put inside this PR. We had a discussion internally about the scope of this PR and issues were raised which prevent us from accepting this PR as it is.

First off, I had to clear up the confusion of why we have our own implementation for handling manifest. Our implementation handles various implementation (or missimplementation) of other tools.

The second aspect is around the lack of user facing tests. The issues addressed by this PR are both very good problems and we feel we should be testing user facing features instead by adding integration tests. We should strive to test the behavior from the user's point of view. A single test could be enough unless there is reason to believe other edge cases would require more test coverage.

We greatly appreciate the effort that was put inside the test coverage of this PR. Unfortunately, the reality for us is we don't have enough knowledge in the test library used. In the long run, it will cause us a lot of pain during maintenance. Instead, as explained above, it would be better to focus on key integration tests that shows, from a user's perspective, the issues are no longer present. It will also help us focus on the important aspect of the change so we can merge the PR quickly.

What I would suggest is focus on one issue at a time starting with an integration test that shows how a user would write a build script that would exerce the unexpected behavior. Both issues have great reproducing samples that could be used as integration test to set the expectation and test coverage. Feel free to take some part of this PR for your next contribution. I think it would be best to close this PR and resume the work as part of another, more focused PR. Feel free to reopen this PR if you want to refocus this PR into one for the two issues.

I'm looking forward to your next contribution.

**Comment by vlsi** [comment]
@lacasseio , thanks for the reply.
Did you just say the PR could be accepted if I remove all the tests and add a single `build.gradle-like` test that fails?

In other words, I can drop all the quickcheck tests.
I can add a `build.gradle-based` test that generates a jar with multi-byte characters in manifest.

> Both issues have great reproducing samples that could be used as integration test to set the expectation and test coverage

I might be wrong, but I think I can't really fix the issues individually.

Just to remind:
1) https://github.com/gradle/gradle/issues/2295 is for "manifest entry order"
2) https://github.com/gradle/gradle/issues/5225 is for "wrong split of multi-byte values"

Currently, Gradle uses Java's implementation `java.util.jar.Manifest`, and neither of the issues can be implemented from Gradle perspective. At least "manifest writer" must be re-implemented (because Java's implementation has both issues: it corrupts the output, and it shuffles the entries).

In other words, the most common end-to-end case "generate JAR file with manifest" would happen to fix two issues at the same time.

Do you suggest to split the PR into "manifest writer", "manifest reader" parts, and drop the tests?

**Comment by lacasseio** [comment]
I think it's more about the iterative process of fixing only what is truly required. It may be true that both issues may need a similar code change but I don't think both issues need to be fixed at the same time. The more we add to a PR the harder it is for us to understand what is fixing what from a PR which makes the review process longer or cause a PR to be closed without fixing the issue. Unfortunately, the core team is often overloaded and the simplest PR the better for us.

To that extend, I took a stab at creating a PR for each issues and it shows some interesting point for each issue.

Issue https://github.com/gradle/gradle/issues/5225, doesn't need the manifest entry order problem to be solved. According to https://github.com/gradle/gradle/pull/11463 it only need a "custom" implementation of the manifest writer that can mostly be pulled from `java.util.jar.Manifest` with. We still convert our Manifest instance to Java Manifest which is the cause for the manifest ordering (more on that later) but we can still get away from that change when fixing this issue. The PR is still going through CI so it may break something and the fix may not be complete but it's pretty close to keeping things simple for us. Most of the "writing" is implementation details so we don't need to do more modeling than that.

As for issue https://github.com/gradle/gradle/issues/2295, it was brought up internally that it should be fixed. The first step is to create an test coverage from the user's perspective to validate the error is present. PR https://github.com/gradle/gradle/pull/11459 do just that. It went through CI and it's showing that ordering is in theory fixed for JDK 11 (or so) + when the internal implementation of `Manifest` started using an `LinkedHashMap`. The questions becomes more about should Gradle fix the ordering for users below JDK 11. Both possibility doesn't break Gradle caching as the map ordering will be consistent for the same JDK. The problem, from Gradle perspective, is when users are mixing JDK version within the same organization using Gradle build cache or, as explained in the issue, to do package sealing. It's good to note that how PR https://github.com/gradle/gradle/pull/11459 isn't exactly how we may want to solve the issue as it's pretty hacky. However, it may be enough if no regression are introduced.

As explained, it's not just about removing tests, it's about making the fix simple and straight to the point.
